compiler/qsc_qasm3/src/symbols.rs

```diff
@@ -0,0 +1,436 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use oq3_semantics::types::{IsConst, Type};
+use qsc::Span;
+use rustc_hash::FxHashMap;
+
+/// We need a symbol table to keep track of the symbols in the program.
+/// The scoping rules for QASM3 are slightly different from Q#. This also
+/// means that we need to keep track of the input and output symbols in the
+/// program. Additionally, we need to keep track of the types of the symbols
+/// in the program for type checking.
+/// Q# and QASM have different type systems, so we track the QASM semantic
+/// type in addition to the corresponding Q# type.
+
+/// A symbol ID is a unique identifier for a symbol in the symbol table.
+/// This is used to look up symbols in the symbol table.
+/// Every symbol in the symbol table has a unique ID.
+#[derive(Debug, Default, Clone, Copy)]
+pub struct SymbolId(pub u32);
+
+impl SymbolId {
+    /// The successor of this ID.
+    #[must_use]
+    pub fn successor(self) -> Self {
+        Self(self.0 + 1)
+    }
+}
+
+impl From<u32> for SymbolId {
+    fn from(val: u32) -> Self {
+        SymbolId(val)
+    }
+}
+
+impl From<SymbolId> for u32 {
+    fn from(id: SymbolId) -> Self {
+        id.0
+    }
+}
+
+impl From<SymbolId> for usize {
+    fn from(value: SymbolId) -> Self {
+        value.0 as usize
+    }
+}
+
+impl From<usize> for SymbolId {
+    fn from(value: usize) -> Self {
+        SymbolId(
+            value.try_into().unwrap_or_else(|_| {
+                panic!("Value, {value}, does not fit into {}", stringify!($id))
```

Comment:
Here `stringify!($id)` will just expand to the literal "$id" since `$id` is not bound inside a macro context. Is that what you meant? In that case you can just write:
```rust
panic!("Value, {value}, does not fit into $id")
```

Reply:
I guess what you really meant is
```rust
panic!("Value, {value}, does not fit into usize")
```
?